crates/pop-wallet-integration/src/lib.rs

```diff
@@ -0,0 +1,161 @@
+use axum::{
+	routing::{get, post},
+	Router,
+};
+use serde::Serialize;
+use std::{path::PathBuf, sync::Arc};
+use tokio::sync::{oneshot, Mutex};
+use tower_http::services::ServeDir;
+
+/// Data to be sent to frontend for signing.
+#[derive(Serialize)]
+pub struct Data {
+	chain_rpc: String,
+	data_type: DataType,
+}
+
+/// The type of transaction with specific data for signing (contract or parachain).
+#[derive(Serialize)]
+pub enum DataType {
+	/// Parachain call, where Vec<u8> is the encoded call data.
+	Parachain(Vec<u8>),
+	Contract(ContractArgs),
+}
+
+/// Contract specific data variations.
+pub mod contracts {
+	use super::Serialize;
+	#[derive(Serialize)]
+	pub struct ContractArgs {
+		call_type: ContractCallType,
+		storage_deposit_limit: Option<String>,
+		/// The binary (wasm, polkavm) of the contract.
+		code: Option<Vec<u8>>,
+	}
+
+	#[derive(Serialize)]
+	pub enum ContractCallType {
+		// no unique fields.
+		Upload,
+		Instantiate(InstantiateArgs),
+		Call(CallArgs),
+	}
+
+	/// Arguments for instantiating a contract.
+	#[derive(Serialize)]
+	pub struct InstantiateArgs {
+		constructor: String,
+		args: Vec<String>,
+		value: String,
+		gas_limit: Option<u64>,
+		proof_size: Option<u64>,
+		salt: Option<Vec<u8>>,
+	}
+
+	/// Arguments for calling a contract.
+	#[derive(Serialize)]
+	pub struct CallArgs {
+		address: String,
+		message: String,
+		args: Vec<String>,
+		value: String,
+		gas_limit: Option<u64>,
+		proof_size: Option<u64>,
+	}
+}
+use crate::contracts::ContractArgs;
+
+struct StateHandler {
+	shutdown_tx: Option<oneshot::Sender<()>>,
+	signed_payload: Option<String>,
+}
+
+/// Manages the wallet integration for secure signing of transactions.
+pub struct WalletIntegrationManager {
+	frontend_path: PathBuf,
+	// cloning can be expensive (e.g. contract code stored in-memory)
+	data: Arc<Data>,
+	signed_payload: Option<String>,
+}
+
+impl WalletIntegrationManager {
+	/// - frontend_path: Path to the wallet-integration frontend.
+	/// - data: Data to be sent to the frontend for signing.
+	pub fn new(frontend_path: PathBuf, data: Data) -> Self {
+		Self { frontend_path, data: Arc::new(data), signed_payload: Default::default() }
+	}
+
+	/// Serves the wallet-integration frontend and an API for the wallet to get
+	/// the necessary data and submit the signed payload.
+	pub async fn run(&mut self) {
+		// used to signal shutdown.
+		let (tx, rx) = oneshot::channel();
+
+		// shared state between routes. Will be used to store the signed payload.
+		let state =
+			Arc::new(Mutex::new(StateHandler { shutdown_tx: Some(tx), signed_payload: None }));
+
+		// will shutdown when the signed payload is received
+		let app = Router::new()
+			// cloning Arcs is cheap
```

Comment:
on line [76](https://github.com/r0gue-io/pop-cli/pull/362/commits/173ec1b707f2fb764921b088fb5502e6cc6919c8#diff-15f82683f5b143fff44ed297f03becc696d7983f7abe1727f94382979a2fff70R76) you mention that cloning the field data can be expensive while being an `Arc` as well.

The `Data` struct should never be cloned in any case, right ?

Reply:
https://github.com/r0gue-io/pop-cli/pull/362/commits/cfe41f00110093a32bd2006efafad443b11d0085